kong/tools/wrpc/message.lua

```diff
@@ -0,0 +1,174 @@
+        end
+
+      else
+        if response_future.raw then
+          response_future:done(payload)
+        else
+          response_future:done(
+            decodearray(pb_decode, rpc.output_type, payload.payloads))
+        end
+      end
+
+    elseif rpc.response_handler then
+      pcall(rpc.response_handler,
+        wrpc_peer, decodearray(pb_decode, rpc.output_type, payload.payloads))
+    end
+
+  else
+    -- incoming method call
+    if rpc.handler then
+      local input_data = decodearray(pb_decode, rpc.input_type, payload.payloads)
+      local ok, output_data = ok_wrapper(
+        pcall(rpc.handler, wrpc_peer, table_unpack(input_data, 1, input_data.n)))
+      if not ok then
+        local err = tostring(output_data[1])
+        ngx_log(ERR, ("[wrpc] Error handling %q method: %q"):format(rpc.name, err))
+
+        send_error(wrpc_peer, {
+          etype = "ERROR_TYPE_UNSPECIFIED",
+          description = err,
+        })
+
+        return nil, err
+      end
+
+      wrpc_peer:send_payload({
+        mtype = "MESSAGE_TYPE_RPC", -- MESSAGE_TYPE_RPC,
+        svc_id = rpc.svc_id,
+        rpc_id = rpc.rpc_id,
+        ack = payload.seq,
+        payload_encoding = "ENCODING_PROTO3",
+        payloads = encodearray(pb_encode, rpc.output_type, output_data),
+      })
+
+    else
+      -- rpc has no handler
+      send_error(wrpc_peer, {
+        etype = "ERROR_TYPE_INVALID_RPC",   -- invalid here, not in the definition
+        description = "Unhandled method",
+      })
+    end
+  end
+end
+
+--- Handle incoming error message (mtype == MESSAGE_TYPE_ERROR).
+function _M.handle_error(wrpc_peer, payload)
+  local etype = payload.error and payload.error.etype or "--"
+  local errdesc = payload.error and payload.error.description or "--"
+  ngx_log(NOTICE, string.format(
+    "[wRPC] Received error message, %s.%s:%s (%s: %q)",
+    payload.svc_id, payload.rpc_id, payload.ack, etype, errdesc
+  ))
+
+  local ack = tonumber(payload.ack) or 0
+  if ack > 0 then
+    -- response to a previous call
+    local response_future = wrpc_peer.responses[ack]
+    if response_future then
+      if response_future.deadline and response_future.deadline < ngx_now() then
+        if response_future.expire then
+          response_future:expire()
+        end
+
+      else
+        assert(response_future.error,
+          "response future does not has a error handler!")
+        response_future:error(etype, errdesc)
+      end
+
+    else
+      ngx_log(ERR, 
+        "reciving error response for a call not initiated by this peer.",
+        " Service ID: ", payload.svc_id, " RPC ID: ", payload.rpc_id)
+      local rpc = wrpc_peer.service:get_rpc(payload.svc_id .. payload.rpc_id)
+      if rpc and rpc.error_handler then
+        local ok, err = pcall(rpc.error_handler, wrpc_peer, etype, errdesc)
+        if not ok then
+          ngx_log(ERR, "error thrown when handling RPC error: ", err)
+        end
+      end
```

Comment:
need an error handling for else of "if rpc and rpc.error_handler", just a log for now is okay

Reply:
done